### PR TITLE
First implementation of spack file for Pflotran

### PR DESCRIPTION
**Pull request by BarrySmith**
Funded-by: IDEAS
Project: IDEAS/xSDK
Time: 2 hour

### CODE REVIEW

**Comment by tgamblin** [comment]
@BarrySmith: FYI, since you asked before, you can click the little gear by `Labels` on the right to add labels.  There's a search box there so you can generally just type `xsdk` in it and hit return.

File: var/spack/repos/builtin/packages/pflotran/package.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+class Pflotran(Package):
+    """PFLOTRAN is an open source, state-of-the-art massively parallel 
+       subsurface flow and reactive transport code.
+    """
+
+    homepage = "http://http://www.pflotran.org"
+
+    version('develop', hg='https://bitbucket.org/pflotran/pflotran-xsdk')
+
+    variant('debug', default=False,
```
Comment (davydden) [inline var/spack/repos/builtin/packages/pflotran/package.py]: this appears to be unused.

Comment (davydden) [inline var/spack/repos/builtin/packages/pflotran/package.py]: this appears to be unused.

File: var/spack/repos/builtin/packages/pflotran/package.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    depends_on('mpi')
+    depends_on('hdf5@1.8.12+mpi+fortran')
+    depends_on('petsc@develop')
+
+    def install(self, spec, prefix):
+        configure('--prefix=%s' % prefix)
+
+        make("all", parallel=False)
+        make("install", parallel=False)
+        make("check", parallel=False)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pflotran/package.py]: If you add `parallel = False` above the `install()` method, it will run all `make()` methods in serial. See http://spack.readthedocs.io/en/latest/packaging_guide.html#parallel-builds for more information.

Comment (tgamblin) [inline var/spack/repos/builtin/packages/pflotran/package.py]: I added this -- will merge when Travis is happy.

**Comment by adamjstewart** [comment]
Might as well switch this to AutotoolsPackage and delete the install method.

File: var/spack/repos/builtin/packages/pflotran/package.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+from spack import *
+
+
+class Pflotran(Package):
+    """PFLOTRAN is an open source, state-of-the-art massively parallel
+       subsurface flow and reactive transport code.
+    """
+
+    homepage = "http://http://www.pflotran.org"
+
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/pflotran/package.py]: This URL looks wrong...

Comment (BarrySmith) [inline var/spack/repos/builtin/packages/pflotran/package.py]: Indeed, how embarrassing. Also removed the unneed install method by switching to AutotoolsPackage @adamjstewart Thanks.

File: var/spack/repos/builtin/packages/pflotran/package.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+       subsurface flow and reactive transport code.
+    """
+
+    homepage = "http://www.pflotran.org"
+
+    version('develop', hg='https://bitbucket.org/pflotran/pflotran-xsdk')
+
+    depends_on('mpi')
+    depends_on('hdf5@1.8.12+mpi+fortran')
+    depends_on('petsc@develop')
```
Comment (BarrySmith) [inline var/spack/repos/builtin/packages/pflotran/package.py]: Yes, requires exactly those versions.

File: var/spack/repos/builtin/packages/pflotran/package.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+    parallel = False
+
```
Comment (BarrySmith) [inline var/spack/repos/builtin/packages/pflotran/package.py]: removed blank line

File: var/spack/repos/builtin/packages/pflotran/package.py (hunk 4 above)
Comment (citibeth) [inline var/spack/repos/builtin/packages/pflotran/package.py]: No, we cannot merge packages that depend on or provide only a `@develop` version; because that circumvents Spack's security.  To see what I mean... when someone says `spack install pflotran`, Spack will be unable to checksum the download from either `pflotran` or `petsc`.  If either project has been compromised, Spack would happily install that compromised code on the user's machine.

This can be corrected by creating a Spack-only "release" for both `petsc` and `pflotran`.  Then you will have an install that you know works predictably.  See here on how to do it:
    https://spack.readthedocs.io/en/latest/workflows.html#upstream-bug-fixes

Basically, you want to add "DIY" versions to both `pflotran` and `petsc` that look something like:
```
# This is not a tagged release of pyproj.
# The changes in this "version" fix some bugs, especially with Python3 use.
version('1.9.5.1.1',
     git='https://github.com/jswhit/pyproj.git',
     commit='0be612cc9f972e38b50a90c946a9b353e2ab140f')
```

**Comment by tgamblin** [comment]
@BarrySmith: does `pflotran` depend on a release version of `PETSc` or does it specifically need something in `develop`?

**Comment by citibeth** [comment]
Even if it needs something  from `develop`... for the purposes of a public
release, I would still advocate for nailing down AT LEAST one version of
`PETSc` and `pflotran` that you know builds.  It's OK to leave the
`develop` versions in there, as long a you have at least one nailed-down
version.

This way:

1. The unsuspecting user who types `spack install pflotran` won't get
unverifiable software.

2. You ensure that `spack install pflotran` will always work; which is not
the case if you leave things on floating versions.

**Comment by tgamblin** [comment]
@citibeth: I agree; although currently we *do* have versions of some packages like this in the repo already.  I think it's good not to proliferate, but we don't have security rules clearly documented or enforced in the contributor guide or in Spack itself.  I was planning to address this after `0.10` by adding some of the preferences we've discussed.  Do you think that is reasonable?

**Comment by citibeth** [comment]
It's my regret that I haven't had the time to work on the improved security
stuff.

> I was planning to address this after 0.10 by adding some of the
preferences we've discussed. Do you think that is reasonable?
Do you mean updating docs?  Yes, I think it's reasonable.

I would still encourage the author to just nail down a version that works.
That means adding one `version()` line to each of `pflotran` and `petsc`.
Not only does it prevent proliferation of insecure packages --- it also
improves the `pflotran` release by documenting that SOME version is known
to work.

**Comment by BarrySmith** [comment]
I have no problem making  "release versions" for these packages as @citibeth requested. 

A suggestion: that when a package only has a develop version Spack should require the user to specifically request the develop version, otherwise stop indicating no valid version available. In other words  

`spack install pflotran`

would not work, only 

`spack install pflotran@develop`

This would prevent the problem @citibeth pointed out by requiring the user to concisously pick the develop version but not require these extraneous extra versions in the spack files. 

Likely there will not be a "real" release of pflotran that uses a "real" release of PETSc for a couple months, but I want to get it into the xSDK testing suite now.

**Comment by citibeth** [comment]
> A suggestion: that when a package only has a develop version Spack should
> require the user to specifically request the develop version, otherwise
> stop indicating no valid version available.
>
Thank you... a more thorough revamp of security has been contemplated.

**Comment by BarrySmith** [comment]
I think I've satisfied all the changes requested for this package so it should be ready to merge.
